src/execution/value.rs

```diff
@@ -261,14 +261,109 @@ pub enum Value {
     F64(F64),
     // F64,
     // V128,
+    Ref(Ref),
 }
 
-#[derive(Clone, Debug, PartialEq)]
-#[allow(dead_code)]
+#[derive(Clone, Copy, Debug, PartialEq)]
 pub enum Ref {
-    Null,
-    // Func,
-    // Extern,
+    Func(FuncAddr),
+    Extern(ExternAddr),
+}
+
+impl Ref {
+    pub fn default_from_ref_type(rref: RefType) -> Self {
+        match rref {
+            RefType::ExternRef => Self::Extern(ExternAddr::default()),
+            RefType::FuncRef => Self::Func(FuncAddr::default()),
+        }
+    }
+
+    pub fn is_null(&self) -> bool {
+        match self {
+            Self::Extern(extern_addr) => extern_addr.addr.is_none(),
+            Self::Func(func_addr) => func_addr.addr.is_none(),
+        }
+    }
+
+    pub fn is_specific_func(&self, func_id: u32) -> bool {
+        match self {
+            Self::Func(func_addr) => func_addr.addr == Some(func_id as usize),
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl Display for Ref {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            Ref::Func(func_addr) => write!(f, "FuncRef({:?})", func_addr),
+            Ref::Extern(extern_addr) => write!(f, "ExternRef({:?})", extern_addr),
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+pub struct FuncAddr {
```

Comment:
Doc String here to reference the wasm tables?
Maybe new types? like `FuncIdx = usize`